provider-service/vai/internal/vai_event_flow.go

```diff
@@ -0,0 +1,257 @@
+package internal
+
+import (
+	aiplatform "cloud.google.com/go/aiplatform/apiv1"
+	"cloud.google.com/go/aiplatform/apiv1/aiplatformpb"
+	"context"
+	"github.com/go-logr/logr"
+	"github.com/googleapis/gax-go/v2"
+	"github.com/sky-uk/kfp-operator/argo/common"
+	. "github.com/sky-uk/kfp-operator/provider-service/base/pkg"
+	"github.com/sky-uk/kfp-operator/provider-service/base/pkg/streams"
+)
+
+const (
+	PushedModelArtifactType        = "tfx.PushedModel"
+	ModelPushedMetadataProperty    = "pushed"
+	ModelPushedMetadataValue       = 1
+	ModelPushedDestinationProperty = "pushed_destination"
+)
+
+var labels = struct {
+	PipelineName              string
+	PipelineNamespace         string
+	PipelineVersion           string
+	RunConfigurationName      string
+	RunConfigurationNamespace string
+	RunName                   string
+	RunNamespace              string
+}{
+	PipelineName:              "pipeline-name",
+	PipelineNamespace:         "pipeline-namespace",
+	PipelineVersion:           "pipeline-version",
+	RunConfigurationName:      "runconfiguration-name",
+	RunConfigurationNamespace: "runconfiguration-namespace",
+	RunName:                   "run-name",
+	RunNamespace:              "run-namespace",
+}
+
+type VaiEventFlow struct {
+	ProviderConfig    VAIProviderConfig
+	PipelineJobClient PipelineJobClient
+	Logger            logr.Logger
+	context           context.Context
+	in                chan StreamMessage[string]
+	out               chan StreamMessage[*common.RunCompletionEventData]
+	errorOut          chan error
+}
+
+func (vef *VaiEventFlow) In() chan<- StreamMessage[string] {
+	return vef.in
+}
+
+func (vef *VaiEventFlow) Out() <-chan StreamMessage[*common.RunCompletionEventData] {
+	return vef.out
+}
+
+func (vef *VaiEventFlow) ErrOut() <-chan error {
+	return vef.errorOut
+}
+
+func (vef *VaiEventFlow) From(outlet streams.Outlet[StreamMessage[string]]) streams.Flow[StreamMessage[string], StreamMessage[*common.RunCompletionEventData], error] {
+	go func() {
+		for message := range outlet.Out() {
+			vef.In() <- message
+		}
+	}()
+	return vef
+}
+
+func (vef *VaiEventFlow) To(inlet streams.Inlet[StreamMessage[*common.RunCompletionEventData]]) {
+	go func() {
+		for message := range vef.out {
+			inlet.In() <- message
+		}
+	}()
+}
+
+func (vef *VaiEventFlow) Error(inlet streams.Inlet[error]) {
+	for errorMessage := range vef.errorOut {
+		inlet.In() <- errorMessage
+	}
+}
+
+func NewVaiEventFlow(ctx context.Context, config *VAIProviderConfig, pipelineJobClient *aiplatform.PipelineClient) streams.Flow[StreamMessage[string], StreamMessage[*common.RunCompletionEventData], error] {
+	logger := common.LoggerFromContext(ctx)
+
+	vaiEventFlow := VaiEventFlow{
+		ProviderConfig:    *config,
+		PipelineJobClient: pipelineJobClient,
+		Logger:            logger,
```

Comment:
```suggestion
```

We should remove logger, its available within the context which is already provided within the struct